### PR TITLE
feat(sp): support nofa use sp

### PR DESCRIPTION
**Pull request by SolenoidWGT**
## Motivation

Support nofa using sequence parallel, if in nofa mode, all-gather/reduce-scatter will work on dim 1 else will work on dim 0.
## Modification

TODO
## BC-breaking (Optional)

TODO

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/model/utils.py [hunk 1]
```diff
@@ -6,8 +6,8 @@
 import torch
 import torch.distributed as dist
 import torch.nn.functional as F
+from internlm_accelerator.amp import custom_bwd, custom_fwd
 from torch import Tensor
-from torch.cuda.amp import custom_bwd, custom_fwd
```
Comment (sunpengsdu) [inline internlm/model/utils.py]: this should be moved to the 910B support PR

File: internlm/model/utils.py [hunk 2]
```diff
@@ -185,24 +186,25 @@ def reduce_scatter_raw(
     process_group: ProcessGroup,
     op=dist.ReduceOp.SUM,
     async_op: bool = False,
+    reduce_dim: int = 0,
```
Comment (huangting4201) [inline internlm/model/utils.py]: 这个参数改成scatter_dim是不是更好一些

File: internlm/model/utils.py [hunk 3]
```diff
@@ -313,7 +315,9 @@ def backward(ctx, grad_output, *args):
             grad_input = grad_input.reshape(*batch_shape, grad_input.shape[-1])
             if process_group is not None:
                 reduce_fn = reduce_scatter_raw if sequence_parallel else all_reduce_raw
-                grad_input, handle_grad_input = reduce_fn(grad_input, process_group, async_op=True)
+                grad_input, handle_grad_input = reduce_fn(
+                    grad_input, process_group, async_op=True, reduce_dim=gather_dim
+                )
```
Comment (huangting4201) [inline internlm/model/utils.py]: 这里还是得分if sequence_parallel来写吧，all_reduce_raw没有reduce_dim这个参数

**Comment by huangting4201** [comment]
1. evaluate时，需要在上下文里更改gather_dim和scatter_dim
2. pp里的get_tensor_shape()也需要更新

File: internlm/model/ops/linear.py [hunk 4]
```diff
@@ -242,18 +243,20 @@ def __init__(
...
         self.process_group = process_group
         self.sequence_parallel = sequence_parallel
+        self.gather_dim = comm_dim
 
-    def forward(self, x, gather_dim=0):
+    def forward(self, x):
         # If self.sequence_parallel is True, we're doing Tensor Parallel with sequence parallelism:
         # we do an all_gather of x before doing the matmul.
         # If not, then the input is already gathered.
+        # print(f"ColumnParallelLinearTorch, x: {x.shape}, gather_dim: {self.gather_dim}", flush=True)
```
Comment (yingtongxiong) [inline internlm/model/ops/linear.py]: 这一行可删？

File: internlm/model/ops/linear.py [hunk 5]
```diff
@@ -324,15 +328,19 @@ def __init__(
...
         self.process_group = process_group
         self.sequence_parallel = sequence_parallel
+        self.reduce_dim = comm_dim
 
     def forward(self, x):
         """
         We're doing Tensor Parallel with sequence parallelism: we do the matmul and then
         a reduce_scatter of the result.
         """
+        # print(f"RowParallelLinearTorch, x: {x.shape}, reduce_dim: {self.reduce_dim}", flush=True)
```
Comment (yingtongxiong) [inline internlm/model/ops/linear.py]: 这一行可删

File: internlm/model/utils.py [hunk 6]
```diff
@@ -54,7 +55,7 @@ def forward(ctx, input_: Tensor, process_group: ProcessGroup) -> Tensor:
     @staticmethod
     def backward(ctx, grad_output: Tensor):
         _ = ctx  # avoid lint warning W0613
-        return grad_output, None
+        return grad_output, None, None
```
Comment (yingtongxiong) [inline internlm/model/utils.py]: 这里多返回了一个None？

File: internlm/model/utils.py (hunk 3 above)
Comment (yingtongxiong) [inline internlm/model/utils.py]: all_reduce_raw似乎没有reduce_dim这个参数，这里可以再double check一下？

File: internlm/model/utils.py [hunk 7]
```diff
@@ -651,7 +658,8 @@ def try_import_RMSNorm():
         from apex.normalization.fused_layer_norm import MixedFusedRMSNorm as RMSNorm
 
         return RMSNorm
-    except ModuleNotFoundError:
+    except ImportError:
+        # except ModuleNotFoundError:
```
Comment (yingtongxiong) [inline internlm/model/utils.py]: 这一行也可删？

File: internlm/model/utils.py [hunk 8]
```diff
@@ -428,7 +433,9 @@ def backward(ctx, grad_output, *args):
             grad_input = grad_input.reshape(*batch_shape, grad_input.shape[-1])
             if process_group is not None:
                 reduce_fn = reduce_scatter_raw if sequence_parallel else all_reduce_raw
-                grad_input, handle_grad_input = reduce_fn(grad_input, process_group, async_op=True)
+                grad_input, handle_grad_input = reduce_fn(
+                    grad_input, process_group, async_op=True, reduce_dim=ctx.gather_dim
```
Comment (yingtongxiong) [inline internlm/model/utils.py]: 这里同上，all_reduce_raw没有reduce_dim这个参数，可能在mtp的时候会有问题

File: internlm/model/utils.py (hunk 1 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: internlm/model/ops/linear.py (hunk 4 above)
Comment (SolenoidWGT) [reply internlm/model/ops/linear.py]: Done

File: internlm/model/ops/linear.py (hunk 5 above)
Comment (SolenoidWGT) [reply internlm/model/ops/linear.py]: Done

File: internlm/model/utils.py (hunk 3 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: internlm/model/utils.py (hunk 8 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: internlm/model/utils.py (hunk 3 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: internlm/model/utils.py (hunk 7 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: internlm/model/utils.py (hunk 2 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: internlm/model/utils.py (hunk 6 above)
Comment (SolenoidWGT) [reply internlm/model/utils.py]: Done

File: configs/7B_sft.py [hunk 9]
```diff
@@ -174,7 +174,7 @@
 """
 parallel = dict(
     zero1=dict(size=-1),
-    tensor=dict(size=1, mode="mtp"),
+    tensor=dict(size=2, mode="isp"),
```
Comment (huangting4201) [inline configs/7B_sft.py:177]: config文件的改动撤回吧

File: internlm/model/utils.py [hunk 10]
```diff
@@ -22,6 +22,19 @@
...
+    """(TODO: If adding new pack_data_mode setting, refactor here)
+    Becase training and evaluation may have different scatter/gather dim,
+    so we need to check which dim will be reduced/gathered on the fly.
+    Returns:
+        int: the dim will be operated on.
+    """
+    if gpc.is_evaluating or gpc.config.model.use_flash_attn is False:
+        return 1
+    else:
+        return 0
```
Comment (huangting4201) [inline internlm/model/utils.py:35]: 这里如果考虑NPU场景的话，可能有问题，当FA:true Packed:false，应该返回的是1，但这里会是0

建议改成这样，腾看看有没有问题：
```python
if gpc.is_evaluating:
        return 1
elif gpc.config.data.use_packed_dataset:
        return 0
else:
        return 1
```
